src/csv.pp.ml

```diff
@@ -120,12 +120,51 @@ type in_channel = {
   lstrip_buffer : Buffer.t -> unit;
   rstrip_substring : Bytes.t -> int -> int -> string;
   rstrip_contents : Buffer.t -> string;
+
+  mutable check_bom : bool;
 }
 
 (*
  * CSV input format parsing
  *)
 
+let bom = 
+    [ [| '\xEF' ; '\xBB'; '\xBF' |]  (* UTF-8 *)
+    ; [| '\xFE' ; '\xFF'  |]         (* UTF-16 BE *)
+    ; [| '\xFF' ; '\xFE'  |]         (* UTF-16 LE *)
+    ]
+
+(** check_bom [ic] [bom] return [None] if the stream does not start with a bom
+    and the number of bytes to ignore otherwise.
+ *)
+let check_bom
+ : in_channel -> (char array) list -> int option 
+ = fun ic bom -> 
+     let matched = 
+         List.fold_left (fun matched bom -> 
```

Comment:
Since you want to bail out as soon as you found a match, `find_map` seems more appropriate.